lib/wasi/src/state/env.rs

```diff
@@ -849,114 +853,78 @@ impl WasiEnv {
 
-        let mut already: HashMap<String, Version> = HashMap::new();
-
-        let mut use_packages = uses.into_iter().collect::<VecDeque<_>>();
-
-        let cmd_wasmer = self
-            .bin_factory
-            .commands
-            .get("/bin/wasmer")
-            .and_then(|cmd| cmd.as_any().downcast_ref::<CmdWasmer>());
-
-        let tasks = self.runtime.task_manager();
-
-        while let Some(use_package) = use_packages.pop_back() {
-            if let Some(package) = cmd_wasmer
-                .as_ref()
-                .and_then(|cmd| tasks.block_on(cmd.get_package(use_package.clone())))
-            {
-                // If its already been added make sure the version is correct
-                let package_name = package.package_name.to_string();
-                if let Some(version) = already.get(&package_name) {
-                    if *version != package.version {
-                        return Err(WasiStateCreationError::WasiInheritError(format!(
-                            "webc package version conflict for {} - {} vs {}",
-                            use_package, version, package.version
-                        )));
-                    }
-                    continue;
-                }
-                already.insert(package_name, package.version.clone());
-
-                // Add the additional dependencies
-                for dependency in package.uses.clone() {
-                    use_packages.push_back(dependency);
-                }
+            if !pkg.commands.is_empty() {
+                let _ = root_fs.create_dir(Path::new("/bin"));
 
-                if let WasiFsRoot::Sandbox(root_fs) = &self.state.fs.root_fs {
-                    // We first need to copy any files in the package over to the temporary file system
-                    if let Some(fs) = package.webc_fs.as_ref() {
-                        root_fs.union(fs);
+                for command in &pkg.commands {
+                    let path = format!("/bin/{}", command.name());
+                    let path = Path::new(path.as_str());
+
+                    // FIXME(Michael-F-Bryan): This is pretty sketchy.
+                    // We should be using some sort of reference-counted
+                    // pointer to some bytes that are either on the heap
+                    // or from a memory-mapped file. However, that's not
+                    // possible here because things like memfs and
+                    // WasiEnv are expecting a Cow<'static, [u8]>. It's
+                    // too hard to refactor those at the moment, and we
+                    // were pulling the same trick before by storing an
+                    // "ownership" object in the BinaryPackageCommand,
+                    // so as long as packages aren't removed from the
+                    // module cache it should be fine.
+                    let atom: &'static [u8] = unsafe { std::mem::transmute(command.atom()) };
```

Comment:
Yeah, we discussed switching from `Cow<'static, [u8]>` to `Bytes` for the `ReadOnlyFile` back when we were refactoring a bunch of stuff, but the concern was that would require all of a command's bytes to be in memory.

I think you decided it'd be better to keep the transmute hack in until the filesystem crate gets refactored because it lets us use memory-mapped files.

Reply:
Ah, I remember now, indeed.
I had a good point... 

I guess we can do a `trait Buffer` , and for mmaped webcs we can wrap the `Container` to read the buffer on demand....

But as before, let's just leave it as is and create a followup issue.

Reply:
@Michael-F-Bryan still need to creat that followup issue, then we can resolve.

Reply:
I've created https://github.com/wasmerio/wasmer/issues/3875 and mentioned it in that big comment.